src/hardware/lpt_dac.cpp

```diff
@@ -0,0 +1,164 @@
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+// NOTE: a lot of this code assumes that the callback is called every emulated
+// millisecond
+
+#include "dosbox.h"
+
+#include "pic.h"
+#include "setup.h"
+#include "support.h"
+
+#include "covox.h"
+#include "disney.h"
+#include "ston1_dac.h"
+
+LptDac::LptDac(const std::string &name, const ChannelFeature num_channels,
+               const uint16_t channel_rate_hz)
+        : dac_name(name)
+{
+	using namespace std::placeholders;
+	const auto audio_callback = std::bind(&LptDac::AudioCallback, this, _1);
+
+	// Setup the mixer callback
+	channel = MIXER_AddChannel(audio_callback,
+	                           channel_rate_hz,
+	                           dac_name.c_str(),
+	                           {ChannelFeature::Sleep,
+	                            ChannelFeature::ReverbSend,
+	                            ChannelFeature::ChorusSend,
+	                            ChannelFeature::DigitalAudio,
+	                            num_channels});
+
+	ms_per_frame = 1000.0 / channel->GetSampleRate();
+
+	// Update our status to indicate we're ready
+	status_reg.error = false;
+	status_reg.busy  = false;
+}
+
+void LptDac::BindHandlers(const io_port_t lpt_port, const io_write_f write_data,
+                          const io_read_f read_status, const io_write_f write_control)
+{
+	// Register port handlers for 8-bit IO
+	data_write_handler.Install(lpt_port, write_data, io_width_t::byte);
+
+	const auto status_port = static_cast<io_port_t>(lpt_port + 1u);
+	status_read_handler.Install(status_port, read_status, io_width_t::byte);
+
+	const auto control_port = static_cast<io_port_t>(lpt_port + 2u);
+	control_write_handler.Install(control_port, write_control, io_width_t::byte);
+}
+
+void LptDac::RenderUpToNow()
+{
+	const auto now = PIC_FullIndex();
+
+	// Wake up the channel and update the last rendered time datum.
+	assert(channel);
+	if (channel->WakeUp()) {
+		last_rendered_ms = now;
+		return;
+	}
+	// Keep rendering until we're current
+	assert(ms_per_frame > 0.0);
+	while (last_rendered_ms < now) {
+		last_rendered_ms += ms_per_frame;
+		render_queue.emplace(Render());
+	}
+}
+
+void LptDac::AudioCallback(const uint16_t requested_frames)
+{
+	assert(channel);
+
+	auto frames_remaining = requested_frames;
+
+	// First, add any frames we've queued since the last callback
+	while (frames_remaining && render_queue.size()) {
+		channel->AddSamples_sfloat(1, &render_queue.front()[0]);
```

Comment:
Oh yeah 😄

Reply:
Pretty soon we'll be float-only API into the mixer 😎